Refactor optimization level selection in LGC

s-perron opened the pull request:
Add a parameter to set the optimization level in LGC.  The command line option will overide that value if used.
    
This will be need to try to implement `VK_PIPELINE_CREATE_DISABLE_OPTIMIZATION_BIT` in
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCreateFlagBits.html.

kuhar reviewed (commented):
Once folks are happy with the PR, could you either apply clang-tidy suggestions that make sense, or mark the lines you don't want to be checked with `// NOLINT`?

kuhar commented on lgc/unittests/interface/OptLevelTest.cpp:
```diff
@@ -0,0 +1,45 @@
+ *
+ **********************************************************************************************************************/
+
+#include "lgc/EnumIterator.h"
```
nit: you can include `llvm/ADT/Sequence.h` for `enum_seq` directly. It may also make sense to mark the `CodeGenOpt::Level` safe for iteration in LLVM so that we don't have to use `force_iteration_on_noniterable_enum`.

kuhar replied:
Or alternatively, we could list all opt levels by hand, since there are only a few we are interested in:
```c++
using namespace CodeGenOpt;
for (auto optLevel : {Level::None, ..., Level::Aggressive})
```

amdvlk-admin commented:
### Test summary for commit 68725d1704fd236954e6d44235a82e008349e67d
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 0a0a4ae4ab062d31fcd68863c35952967a8060ee</li>
<li>SPVGEN: dev 6c2f36bcf0f8ff7ba8d20c8b789ca6346d1e16b1</li>
<li>XGL: dev c41a825d1ea17c108e8fb123f082357f849e7823</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228192)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 33955/57048 (59.5%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23093/57048 (40.5%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33797/57048 (59.2%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23250/57048 (40.8%)</li><li>Warnings: 1/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 34085/57048 (59.7%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 22963/57048 (40.3%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35882/57048 (62.9%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 21166/57048 (37.1%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 5f80e83941847de85a4151cdfd4ef02be9915598
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 0a0a4ae4ab062d31fcd68863c35952967a8060ee</li>
<li>SPVGEN: dev 6c2f36bcf0f8ff7ba8d20c8b789ca6346d1e16b1</li>
<li>XGL: dev c41a825d1ea17c108e8fb123f082357f849e7823</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228192)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 33955/57048 (59.5%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23093/57048 (40.5%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33797/57048 (59.2%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23250/57048 (40.8%)</li><li>Warnings: 1/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 34085/57048 (59.7%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 22963/57048 (40.3%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35882/57048 (62.9%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 21166/57048 (37.1%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>


</ul></details>

kuhar commented on lgc/unittests/interface/OptLevelTest.cpp:
```diff
@@ -0,0 +1,45 @@
+ *
+ **********************************************************************************************************************/
+
+#include "lgc/EnumIterator.h"
```
nit: unused include

kuhar reviewed (dismissed):
LGTM but it'd be better if someone from AMD also looked at this.

amdvlk-admin commented:
### Test summary for commit 1b6e2b8b232bae6aef7bcb81f0016340c99a92de
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 0a0a4ae4ab062d31fcd68863c35952967a8060ee</li>
<li>SPVGEN: dev 6c2f36bcf0f8ff7ba8d20c8b789ca6346d1e16b1</li>
<li>XGL: dev c41a825d1ea17c108e8fb123f082357f849e7823</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228192)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 33955/57048 (59.5%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23093/57048 (40.5%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33797/57048 (59.2%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23250/57048 (40.8%)</li><li>Warnings: 1/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 34085/57048 (59.7%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 22963/57048 (40.3%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35882/57048 (62.9%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 21166/57048 (37.1%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>


</ul></details>

kuhar commented on lgc/include/lgc/patch/Patch.h:
```diff
@@ -73,7 +73,7 @@ class LegacyPatchCheckShaderCache;
 //
 // @param passRegistry : Pass registry
-inline static void initializePatchPasses(llvm::PassRegistry &passRegistry) {
+inline static void initializePatchPasses(llvm::PassRegistry &passRegistry) { // NOLINT
```
The lint here is weird; the comments don't seem to silence it for whatever reason. I think it may need something like [`[[gnu::used]]`](https://clang.llvm.org/docs/AttributeReference.html#used), but I'd just ignore it instead

kuhar replied:
Ah, I know why. This seems to be because this is a free function declared as `static`. I think removing `static` will resolve this.

amdvlk-admin commented:
### Test summary for commit 93466e1f2ad3d3773a469e0ba6a10b702276584c
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 0a0a4ae4ab062d31fcd68863c35952967a8060ee</li>
<li>SPVGEN: dev 6c2f36bcf0f8ff7ba8d20c8b789ca6346d1e16b1</li>
<li>XGL: dev c41a825d1ea17c108e8fb123f082357f849e7823</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228192)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 33955/57048 (59.5%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23093/57048 (40.5%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33797/57048 (59.2%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23250/57048 (40.8%)</li><li>Warnings: 1/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 34085/57048 (59.7%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 22963/57048 (40.3%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35882/57048 (62.9%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 21166/57048 (37.1%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 543eb181b732292311a020c1c6c8f4f86b073b9f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 0a0a4ae4ab062d31fcd68863c35952967a8060ee</li>
<li>SPVGEN: dev 6c2f36bcf0f8ff7ba8d20c8b789ca6346d1e16b1</li>
<li>XGL: dev c41a825d1ea17c108e8fb123f082357f849e7823</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/228192)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 33955/57048 (59.5%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23093/57048 (40.5%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33796/57048 (59.2%)</li>
<li><details><summary>Failed: 1/57048 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.pipeline_barrier.all.1048576_vertex_buffer_stride_4
Stack trace: Script:
10:HostMemoryAccess Result differs from reference, Expected: 0x0000002e, Got: 0x00000017, At offset: 65111

```
</p></details></li>

<li>Not Supported: 23250/57048 (40.8%)</li><li>Warnings: 1/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 34085/57048 (59.7%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 22963/57048 (40.3%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35882/57048 (62.9%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 21166/57048 (37.1%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>


</ul></details>

bsaleil commented on lgc/patch/Patch.cpp:
```diff
@@ -206,7 +191,8 @@ void Patch::addPasses(PipelineState *pipelineState, lgc::PassManager &passMgr, b
 void LegacyPatch::addPasses(PipelineState *pipelineState, legacy::PassManager &passMgr, ModulePass *replayerPass,
-                            Timer *patchTimer, Timer *optTimer, Pipeline::CheckShaderCacheFunc checkShaderCacheFunc)
+                            Timer *patchTimer, Timer *optTimer, Pipeline::CheckShaderCacheFunc checkShaderCacheFunc,
+                            CodeGenOpt::Level optLevel)
```
Could you add the `optLevel` parameter to the comment above ? (And maybe also the other missing parameters ?)

bsaleil commented on lgc/patch/Patch.cpp:
```diff
@@ -343,8 +329,8 @@ void LegacyPatch::addPasses(PipelineState *pipelineState, legacy::PassManager &p
 // @param [in/out] passMgr : Pass manager to add passes to
-void Patch::addOptimizationPasses(lgc::PassManager &passMgr) {
-  LLPC_OUTS("PassManager optimization level = " << cl::OptLevel << "\n");
+void Patch::addOptimizationPasses(lgc::PassManager &passMgr, CodeGenOpt::Level optLevel) {
```
Same here, add `optLevel` to the comment.

bsaleil commented on lgc/patch/Patch.cpp:
```diff
@@ -397,8 +383,8 @@ void Patch::addOptimizationPasses(lgc::PassManager &passMgr) {
 // @param [in/out] passMgr : Pass manager to add passes to
-void LegacyPatch::addOptimizationPasses(legacy::PassManager &passMgr) {
-  LLPC_OUTS("PassManager optimization level = " << cl::OptLevel << "\n");
+void LegacyPatch::addOptimizationPasses(legacy::PassManager &passMgr, CodeGenOpt::Level optLevel) {
```
Same here, add `optLevel` to the comment.

bsaleil commented on lgc/state/LgcContext.cpp:
```diff
@@ -218,8 +214,9 @@ bool LgcContext::isGpuNameValid(llvm::StringRef gpuName) {
-LgcContext *LgcContext::Create(LLVMContext &context, StringRef gpuName, unsigned palAbiVersion) {
-  assert(Initialized && "Must call LgcContext::Initialize before LgcContext::Create");
+LgcContext *LgcContext::create(LLVMContext &context, StringRef gpuName, unsigned palAbiVersion,
+                               CodeGenOpt::Level optLevel) {
```
Same here, add `optLevel` to the comment.

bsaleil reviewed (commented):
LGTM with minor comments

s-perron commented:
V2: Updated the comments as @bsaleil  suggested, and removed `static` from `initializePatchPasses`, as @kuhar suggested, to see if that fixes the lint issue.  I don't think it is needed.  The "inline" makes the compiler generate a weak definition.  As long at the one definition rule is not broken, all the static does is bloat the exe size.

amdvlk-admin commented:
### Test summary for commit b5cca331cb9eabe4d9611609d3f697373a850a41
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev b638e90ca4e6e5a6fc4f00029d62f8e064aa18eb</li>
<li>SPVGEN: dev 6c2f36bcf0f8ff7ba8d20c8b789ca6346d1e16b1</li>
<li>XGL: dev 89a4cf115b9a2972ff00e2fe4959f4f1c8e7175b</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/266476)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 37645/66619 (56.5%)</li>
<li>Failed: 0/66619 (0.0%)</li>
<li>Not Supported: 28974/66619 (43.5%)</li><li>Warnings: 0/66619 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37475/66619 (56.3%)</li>
<li>Failed: 0/66619 (0.0%)</li>
<li>Not Supported: 29143/66619 (43.7%)</li><li>Warnings: 1/66619 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 38643/66619 (58.0%)</li>
<li>Failed: 0/66619 (0.0%)</li>
<li>Not Supported: 27976/66619 (42.0%)</li><li>Warnings: 0/66619 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 41269/66619 (61.9%)</li>
<li>Failed: 0/66619 (0.0%)</li>
<li>Not Supported: 25350/66619 (38.1%)</li><li>Warnings: 0/66619 (0.0%)</li>
</ul></details>


</ul></details>

s-perron commented:
V3: Rebase

s-perron commented:
@amdrexu Can approve again? I had to rebase because of a conflict, and the review was dismissed.

trenouf reviewed (approved):
Re-approving after rebase.

amdvlk-admin commented:
### Test summary for commit b8ac89879ea7ac6bea9528822d297ffced05cd63
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev b638e90ca4e6e5a6fc4f00029d62f8e064aa18eb</li>
<li>SPVGEN: dev 6c2f36bcf0f8ff7ba8d20c8b789ca6346d1e16b1</li>
<li>XGL: dev 89a4cf115b9a2972ff00e2fe4959f4f1c8e7175b</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 3/253346)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 37643/66619 (56.5%)</li>
<li><details><summary>Failed: 2/66619 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.pipeline_barrier.host_read_host_write.65536
Stack trace: Script:
9:HostMemoryAccess Result differs from reference, Expected: 0x0000007e, Got: 0x00000000, At offset: 46464


FAILURE: dEQP-VK.memory.pipeline_barrier.transfer_src_transfer_dst.65536
Stack trace: Script:
11:BufferCopyToBuffer Result differs from reference, Expected: 0x00000014, Got: 0x00000000, At offset: 16384

```
</p></details></li>

<li>Not Supported: 28974/66619 (43.5%)</li><li>Warnings: 0/66619 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33224/53489 (62.1%)</li>
<li><details><summary>Failed: 1/53489 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.clustered.graphics.subgroupclusteredmin_u64vec2
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 20263/53489 (37.9%)</li><li>Warnings: 1/53489 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 38643/66619 (58.0%)</li>
<li>Failed: 0/66619 (0.0%)</li>
<li>Not Supported: 27976/66619 (42.0%)</li><li>Warnings: 0/66619 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 41269/66619 (61.9%)</li>
<li>Failed: 0/66619 (0.0%)</li>
<li>Not Supported: 25350/66619 (38.1%)</li><li>Warnings: 0/66619 (0.0%)</li>
</ul></details>


</ul></details>

s-perron commented:
retest this please

trenouf commented:
The latest test run failed in jenkins, apparently because of the new llvm-project #include <map> thing that Sebastian fixed in another PR. I'll merge this anyway.
